### PR TITLE
Feature/gbi 821 refund pegout

### PR DESCRIPTION
**Pull request by coolwather**

### CODE REVIEW

File: contracts/LiquidityBridgeContract.sol [hunk 1]
```diff
@@ -418,6 +461,12 @@ contract LiquidityBridgeContract {
 
         return keccak256(encodeQuote(quote));
     }
+
+    function validateAndHashPegOutQuote(PegOutQuote memory quote) private view returns (bytes32) {
+        require(address(this) == quote.lbcAddress, "Wrong LBC address");
+
```
Comment (augustoIOV) [inline contracts/LiquidityBridgeContract.sol:467]: shouldn't we also include here this require, as in validateAndHashQuote?
`require(address(bridge) != quote.contractAddress, "Bridge is not an accepted contract address");`

File: test/basic.tests.js [hunk 2]
```diff
@@ -592,79 +587,31 @@ contract('LiquidityBridgeContract', async accounts => {
         expect(contractBalanceBefore.toString()).to.be.eq(contractBalanceAfter.toString());
     });
 
-    it('should verify signature for pegout and fail hash because Bridge address is not valid as destination', async () => {
```
Comment (augustoIOV) [inline test/basic.tests.js:595]: why removing these 2 tests?

File: test/basic.tests.js [hunk 3]
```diff
@@ -675,25 +622,69 @@ contract('LiquidityBridgeContract', async accounts => {
...
 
-        const quoteHash = await instance.hashQuote(utils.asArray(quote));
+        console.log('Value:', quote.valueToTransfer.toString());
+        console.log('Fee:', quote.fee.toString());
+
+        const quoteHash = await instance.hashPegoutQuote(utils.asArray(quote));
         const signature = await web3.eth.sign(quoteHash, liquidityProviderRskAddress);
 
-        const pegoutCall = instance.registerPegOut.call(utils.asArray(quote), signature, {value: quote.val}); // needs to add call fee as well
+        const pegoutCall = instance.registerPegOut.call(utils.asArray(quote), signature, {value: web3.utils.toBN(0.5)}); // needs to add call fee as well
```
Comment (augustoIOV) [inline test/basic.tests.js:640]: remove the comment?

File: test/basic.tests.js [hunk 4]
```diff
@@ -675,25 +622,69 @@ contract('LiquidityBridgeContract', async accounts => {
...
         await truffleAssertions.reverts(pegoutCall, "LBC: msg value doesnt match quote");
 
         const balanceAfter = await web3.eth.getBalance(instance.address);
         expect(balanceBefore).to.be.eq(balanceAfter);
     })
+
+    // it.only('Should refundPegOut', async () => {
```
Comment (augustoIOV) [inline test/basic.tests.js:647]: why is this commented?

File: test/basic.tests.js (hunk 2 above)
Comment (coolwather) [reply test/basic.tests.js:595]: Was removed because they are related to the bridge address.

File: test/basic.tests.js (hunk 3 above)
Comment (coolwather) [reply test/basic.tests.js:640]: Fixed

File: contracts/LiquidityBridgeContract.sol [hunk 5]
```diff
@@ -381,23 +400,43 @@ contract LiquidityBridgeContract {
...
+        bytes32 derivationAddress
+    ) public noReentrancy {
+        bytes32 quoteHash = validateAndHashPegOutQuote(quote);
+        require(processedPegOutQuotes[quoteHash] == 2, "LBC: Quote not processed");
+        require(quoteHash == derivationAddress, "LBC: Invalid quote hash");
+        require(block.timestamp <= quote.expireDate, "LBC: Quote expired");
+        require(block.number <= quote.expireBlocks, "LBC: Quote expired");
+        require(msg.sender == quote.lpAddress, "LBC: Wrong sender");
+        require(bridge.getBtcTransactionConfirmations(btcTxHash, btcBlockHeaderHash, partialMerkleTree, merkleBranchHashes) >= quote.transferConfirmations, "LBC: Don't have required confirmations");
+        payable(address(uint160(uint256(derivationAddress)))).transfer(quote.valueToTransfer);
```
Comment (augustoIOV) [inline contracts/LiquidityBridgeContract.sol:437]: who is getting the refund here?

**Review by joaoiovlabs — APPROVED** [review]

File: contracts/LiquidityBridgeContract.sol [hunk 6]
```diff
@@ -381,23 +401,41 @@ contract LiquidityBridgeContract {
...
+    ) public noReentrancy {
+        bytes32 quoteHash = validateAndHashPegOutQuote(quote);
+        require(processedPegOutQuotes[quoteHash] == 2, "LBC: Quote not processed");
+        require(block.timestamp <= quote.expireDate, "LBC: Quote expired by date");
+        require(block.number <= quote.expireBlocks, "LBC: Quote expired by blocks");
+        require(msg.sender == quote.lpAddress, "LBC: Wrong sender");
+        require(bridge.getBtcTransactionConfirmations(btcTxHash, btcBlockHeaderHash, partialMerkleTree, merkleBranchHashes) >= quote.transferConfirmations, "LBC: Don't have required confirmations");
+        payable(address(uint160(uint256(quote.derivationAddress)))).transfer(quote.valueToTransfer);
+        decreasePegOutBalance(quote.rskRefundAddress, quote.valueToTransfer);
+        delete processedPegOutQuotes[quoteHash];
```
Comment (manfioLP) [inline contracts/LiquidityBridgeContract.sol:438]: shouldn't we keep track of this somehow?

**Review by manfioLP — COMMENTED** [review]
just curious if we shouldn't store the removed data somewhere

File: contracts/LiquidityBridgeContract.sol (hunk 6 above)
Comment (coolwather) [reply contracts/LiquidityBridgeContract.sol:438]: It's tracked on the blockchain, you always can look at historical data and saw that value. Create another struct to track this information will cause an increase in the gas of the transaction.

Comment (manfioLP) [reply contracts/LiquidityBridgeContract.sol:438]: you're right. good point.

**Review by manfioLP — APPROVED** [review]
LGTM

**Pull request merged**